rfd/0085-better-react-testing-remove-default-exports.md

```diff
@@ -0,0 +1,165 @@
+---
+authors: Ryan Clark (ryan.clark@goteleport.com)
+state: draft
+---
+
+# RFD 85 - Better React testing, remove default exports
+
+## What
+
+Improve the webapps codebase by changing the way we test our components and write our component stories.
+
+## Why
+
+The current convention is to use Storybook to render our large, full components in their different states.  This has a
+couple of knock-on effects.
+
+### Component Architecture
+
+As it's hard to mock data in Storybook, we end up exporting a component that is suitable for rendering in Storybook,
+such as:
+
+```typescript jsx
+export function DesktopSession(props: State) {
+  const {
+    directorySharingState,
+    setDirectorySharingState,
+    clipboardState,
+    fetchAttempt,
+    tdpConnection,
+    disconnected,
+    wsConnection,
+    setTdpConnection,
+  } = props;
+  // etc
+
+  return (
+    <Flex flexDirection="column">
+      // etc
+    </Flex>
+  );
+}
+```
+
+This component takes all of its state and business logic from the props passed in.
+
+For use in the actual application, we wrap the component above in a default export which populates the props the
+component needs by using a hook.
+
+```typescript jsx
+export default function Container() {
+  const state = useDesktopSession();
+  return <DesktopSession {...state} />;
+}
+```
+
+
+### Testing
+
+As a result of the above, we end up testing the `DesktopSession` component instead of the actual component (`Container`) that is
+rendered in the UI.
+
+A story may look like this:
+
+```typescript jsx
+export const FetchError = () => (
+  <DesktopSession
+    {...props}
+    fetchAttempt={{ status: 'failed', statusText: 'some fetch  error' }}
+    tdpConnection={{ status: 'success' }}
+    wsConnection={'open'}
+    disconnected={false}
+  />
+);
+```
+
+Which is then tested like so:
+
+```typescript jsx
+test('fetch error', () => {
+  const { getByTestId } = render(<FetchError />);
+  expect(getByTestId('Modal')).toMatchSnapshot();
+});
+```
+
+This results in all the logic in `useDesktopSession` going untested. We're only testing the UI part of the application,
+but skipping all of the business logic.
+
+Storybook isn't best suited for the way we are using it. As a result of testing our large components primarily via stories, we have
+ended up only testing our UI and nothing else.
+
+## Details
+
+### Changing how we write stories
+
+Experience tells us that Stories in Storybook are used most effectively for documenting how to use reusable components. We should therefore
+restrict usage of Storybook mostly to our shared component library, and test our fullscreen components another way.
+
+It's still okay to create our larger components in Storybook (if the developer wants to) but we should not be using these stories in our Jest tests.
+
+It might be worthwhile looking into something like [Chromatic](https://www.chromatic.com/) for visual UI testing.
```

Comment:
I think we should decide here whether or not this is worth integrating. 

It makes sense to me that we have both "visual" and "behavioral" tests, as those two categories comprise the full user experience. With our current testing standard, we have a rough approximation of visual tests using snapshots, and minimal behavioral tests. If we move away from snapshots and focus exclusively on behavioral tests, we seem at risk of allowing the visual details of the app degrade over time (particularly for edge cases). Chromatic appears to solve many of the valid issues raised in the argument cited against snapshot tests ([link](https://medium.com/@sapegin/whats-wrong-with-snapshot-tests-37fbe20dfe8e)). See an example of how Chromatic works [here](https://www.youtube.com/watch?v=9o6uB1X-LZ8).

This argument contradicts what we have below

> By writing tests as described above, snapshot tests typically become redundant and therefore unnecessary.

so we should resolve that discrepancy.

Reply:
> By writing tests as described above, snapshot tests typically become redundant and therefore unnecessary

Just want to provide some clarification. The stories with different states is the way to document the behavior. The snapshot tests of _static_ views is to test against view design/css changes as we go through redesigns/theming (css bugs are by far _the most common_). It's cheaper than writing e2e tests or have each step listed in figma.  The component composition/structure is to make it as simple as possible (1-liner to snapshot a static view). There is a way to automate it with `chromatic` or similar services (they manage snapshots/tests of your stories) but adding another vendor to the build process may actually slow you down by introducing another set of problems.

The reason why we are missing behavior tests here and there is because we had to move fast (and because behavior is very basic for many component). Now If time allows please start adding behavior tests against `containers` or against high level hooks (the structure allows it) but keep the stories and their static snapshots where appropriate as both are for completely different purposes.